### ISSUE TITLE
Filtering by `exists` doesn't result in NonNullable type of field.

### ISSUE THREAD

**Issue by izakfilmalter** [edgedb/edgedb-js#400]
**Code**

```typescript
const query = e.select(e.User, (user) => ({
  filter: e.op('exists', user.name),
  name: true,
}))

type ResultType = $infer<typeof query>
// type ResultType = {
//   name: string | null
// }[]
```

**Schema**

Your application schema. 

```
module default {
  type User {
    property name -> str;
  }
}
```

**Generated EdgeQL**

```
WITH
  __scope_0_User := DETACHED default::User
SELECT __scope_0_User {
  name
}
FILTER (exists __scope_0_User.name)
```

**Error or desired behavior**

I expect to get the following type:
```ts
type ResultType = {
  name: string
}[]
```

**Versions (please complete the following information):**
 - OS: macOS Monterey
 - EdgeDB version:  2.0
 - `edgedb-js` version: 0.21.1

**Comment by izakfilmalter** [comment]
Seems like the same thing is happening for link properties.

```ts
(user) => ({
    '@status': e.op(
      user['@status'],
      'if',
      e.op('exists', user['@status']),
      'else',
      e.Status.pending,
    )
})
```

Doing this still results in `@status` being nullable.

**Comment by colinhacks** [comment]
As discussed on Discord:

I don't think it's possible to fix this in the near future. It's really tricky to figure out how the filter clause should impact the inferred type of something inside the shape in the general case. It's hard enough to do with simple cardinality inference on `filter`.

You can sort of do this with a computed property and assert_exists, though you need to use a key name other than name since they query builder doesn't let you override existing property/link cardinalities

```ts
const query = e.select(e.Contact, (c) => ({
  _name: e.assert_exists(c.name),
  filter: e.op('exists', c.name)
}));
```

In general, `e.assert_exists()` and `e.assert_single()` are the official ways to outsmart the query builder if type inference isn't perfect.

**Comment by izakfilmalter** [comment]
Ya that's fine. I think I'm gonna default value more fields then. I was struggling with doing that strange option thing as well since nested object is hard.

**Comment by datner** [comment]
oh man, I nearly lost my mind with this. @colinhacks now with v1 released, could this be revisited? or is this still too much effort for too little value?

**Comment by raddevon** [comment]
Pinging @jaclarke since he's taken over maintenance on the JS client. I think some people are taking time off too, so he might not get back to you right away, @datner. I'm not sure the status of this myself.

**Comment by jaclarke** [comment]
I think it's unlikely that this will be fixed any time soon. Even the fairly basic cardinality inference that we did do based on the filter expression, we decided to remove in edgedb-js v1.0. This was partly because of performance problems it caused and that it really only worked for a limited number of specific filter expressions, and I think trying to do nullabilty inference from the filter expression would suffer from the same problems.
